onnxruntime/contrib_ops/cpu/quantization/moe_quantization_cpu.cc

```diff
@@ -0,0 +1,407 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#include "contrib_ops/cpu/quantization/moe_quantization_cpu.h"
+#include "core/framework/allocator.h"
+#include "core/framework/buffer_deleter.h"
+#include "core/mlas/inc/mlas.h"
+#include "core/mlas/inc/mlas_q4.h"
+#include "core/mlas/inc/mlas_qnbit.h"
+#include "core/platform/threadpool.h"
+#include "contrib_ops/cpu/moe/moe_utils.h"
+#include <algorithm>
+
+using namespace onnxruntime::common;
+using namespace ONNX_NAMESPACE;
+
+namespace onnxruntime {
+namespace contrib {
+
+#define REGISTER_KERNEL()                                                                  \
+  ONNX_OPERATOR_KERNEL_EX(QMoE, kMSDomain, 1, kCpuExecutionProvider,                       \
+                          (*KernelDefBuilder::Create())                                    \
+                              .MayInplace(0, 0)                                            \
+                              .TypeConstraint("T", BuildKernelDefConstraints<MLFloat16>()) \
+                              .TypeConstraint("T1", BuildKernelDefConstraints<uint8_t>())  \
+                              .TypeConstraint("T2", BuildKernelDefConstraints<float>()),   \
```

Comment:
Suggest to use MLFloat16 (same as T) if perf and accuracy no significant difference. That will be consistent with cuda.